lib/Hook.js

```diff
@@ -0,0 +1,152 @@
+class Hook {
+	constructor(args) {
+		if(!Array.isArray(args)) args = [];
+		this.optionKeys = ["args", "tap", "type"];
+		this._args = args;
+		this.taps = [];
+		this.inspectors = [];
+		this.call = this._call = this.createCompileDelegate("call", "sync", args);
+		this.promise = this._promise = this.createCompileDelegate("promise", "promise", args);
+		this.callAsync = this._callAsync = this.createCompileDelegate("callAsync", "async", args.concat("_callback"));
+		this._x = undefined;
+	}
+
+	compile(options) {
+		const source = this.template(options);
+		return new Function(`return (${source});`)();
+		/*const optionKeys = this.optionKeys;
+		let key = "/";
+		for(let i = 0; i < optionKeys.length; i++)
+			key = key + options[optionKeys[i]] + "/";
+		return new Function(`return (${source});\n//# sourceURL=tapable:///${key}`)();*/
+	}
+
+	getTapType() {
+		if(this.inspectors.length > 0) return "inspect";
+		if(this.taps.length === 0) return "none";
+		if(this.taps.length === 1) {
+			const tap = this.taps[0];
+			return tap.type;
+		}
+		let type = this.taps[0].type;
+		for(let i = 1; i < this.taps.length; i++) {
+			const tap = this.taps[i];
+			if(tap.type !== type) return "multiple";
+		}
+		return "multiple-" + type;
+	}
+
+	createCall(type) {
+		const tap = this.getTapType();
+		if(tap === "sync")
+			this._x = this.taps[0].fn;
+		else if(tap === "async")
+			this._x = this.taps[0].fn;
+		else if(tap === "promise")
+			this._x = this.taps[0].fn;
+		else if(tap === "inspect")
+			this._x = this.taps;
+		else
+			this._x = this.taps.map(t => t.fn);
+		return this.compile({
+			args: this._args,
+			tap: tap,
+			type: type
+		});
+	}
+
+	createCompileDelegate(name, type, args) {
+		return new Function(args, `
+			this.${name} = this.createCall(${JSON.stringify(type)});
+			return this.${name}(${args.join(", ")});
+		`);
+	}
+
+	tap(options, fn) {
+		if(typeof options === "string")
+			options = { name: options };
+		if(typeof options !== "object" || options === null)
+			throw new Error("Invalid arguments to tap(options: Object, fn: function)");
+		options = Object.assign({ type: "sync", fn: fn }, options);
+		if(typeof options.name !== "string" || options.name === "")
+			throw new Error("Missing name for tap");
+		this._insert(options);
+	}
+
+	tapAsync(options, fn) {
```

Comment:
Possible to merge `tapAsync` && `tapPromise` into e.g => `tapP` and handle convert/handle callbacks internally ?

```
Plugin.hooks.hook.tap() (Sync)
Plugin.hooks.hook.tapP() (Async)
```

Not sure if possible though 😛